src/Annotation/AnnotationBuilder.php

```diff
@@ -112,18 +114,16 @@ public function setFormFactory(Factory $formFactory)
     /**
      * Set annotation manager to use when building form from annotations
      *
-     * @param  AnnotationManager $annotationManager
-     * @return $this
+     * @param      object $annotationManager
+     * @deprecated 2.17.0 with laminas-code 4.0 an annotation manager is not available anymore
+     * @return     $this
      */
-    public function setAnnotationManager(AnnotationManager $annotationManager)
+    public function setAnnotationManager(object $annotationManager)
     {
-        $parser = $this->getAnnotationParser();
-        foreach ($this->defaultAnnotations as $annotationName) {
-            $class = __NAMESPACE__ . '\\' . $annotationName;
-            $parser->registerAnnotation($class);
-        }
-        $annotationManager->attach($parser);
-        $this->annotationManager = $annotationManager;
+        trigger_error(sprintf(
+            'Usage of %s is deprecated since v2.17.0, as laminas-code 4.0 does not have an annotation manager anymore.',
```

Comment:
@Ocramius It seems to me that having a `v2.17` release with this change is both a BC-Break and a pita, so I'd like to go all-in and jump directly into a `v3` release with all this removed in favor of `doctrine/annotations`: wdyt?
/cc @froschdesign

Reply:
To me, yes, adding a warning/deprecation is problematic.

Not my call though, as I'm not involved enough in `laminas-form` to understand the next milestones.

Reply:
The warning / deprecation notice would be okay if the feature is still there, then the user can suppress the notice and everything works like before.
But at this point the notice comes too late therefore we must go straight to the next step.

Reply:
It might still be reasonable to release a 2.17.0 (or 2.16.4) with the deprecation notice (and the old behaviour, i.e. laminas-code 3.5.0) and then in parallel make a 3.0 release with all the new stuff. I can provide two PRs if you like. Please let me know if the deprecation notice should be based on 2.16.x or 2.17.x branch.

Reply:
We can't because the class isn't `final` and so we cannot avoid the BC-Break.

Btw all the blame should go to @Ocramius who released a `laminas-code` new major with a tenth of the functionalities of the previous one, and so I'm therefore allowed to do big jumps here too :tongue:

Reply:
"Soft" dependencies, y'all - they come from an age where we were very much unaware of their weight.